### ISSUE TITLE
Fractional scaling support

### ISSUE THREAD

**Issue by CreamyCookie** [mate-desktop/mate-desktop#326]
I'm using Ubuntu Mate 18.04 and I'm pretty happy with it, despite some issues here and there. One problem I have is that Mate's HiDPI scaling seems to be pretty binary. Either scale everything 200% or don't. And that means the text and buttons are gigantic, which looks a bit jarring on a 1440p screen:

![screenshot](https://user-images.githubusercontent.com/3063858/41808380-cbf547f0-76dc-11e8-9a1c-fc48e0190ce6.png)

It also leaves me with a lot less space then I had on Unity. I'd love to scale by 1.5 or 1.75.

**Comment by lukefromdc** [comment]
I think that's something GNOME and GTK devs are just starting to play with. As of GTK 3.22.30, in Gtkinspector you only see integer window scaling. GTK4 (still in development) has so many changes it would be very difficult to ever port MATE to it, and we would be back on the treadmill GTK3 used to be notorious for. 

You can however reset the size of a lot of widgets by  opening mate-appearance-properties, going to fonts->details, and there you will see a  Resolution "dots per inch" spinbox and  and checkbox to enable or disable automatic detection. Changing this will change the size of all text and all widgets containing text. It will not rescale icons or images however.

**Comment by CreamyCookie** [comment]
@lukefromdc Thanks, that improved it a lot. I guess this could be closed then.

**Comment by lukefromdc** [comment]
Yeah we can close this. Fractional scaling in GNOME appears to be a WIP, and as of late last year worked only in Wayland and was considered experimental.  Cinnamon has stated that they cannot implement fractional scaling at this time either:
https://github.com/linuxmint/Cinnamon/issues/3455
A fractional scaling option in GNOME 3.26 (3.28 is now out) was "hidden, experimental and only works with Wayland"

**Comment by lglapinski** [comment]
Although I don't know implementation details but looks like now that Cinnamon has fractional scaling for x11 maybe it is good opportunity to revisit this issue?

**Comment by harrego** [comment]
> maybe it is good opportunity to revisit this issue?

Agreed. GNOME is at a good place with fractional scaling on X11 too, I have no issues with it.

**Comment by lukefromdc** [comment]
Reopening, as GTK now has decent support for this. Now we need to look at how GNOME and Cinnamon did this, presumably this would be something for the xsettings plugin in mate-settings-daemon to handle as it now does integer scaling?

**Comment by juehang** [comment]
A quick look seems to suggest mutter (of Gnome) uses xrandr in combination with integer scaling to produce fractional scaling. I would have expected some blurriness to ensue, but perhaps that is not noticeable with modern HiDPI screens.

https://gitlab.gnome.org/3v1n0/mutter/-/commits/xrandr-scaling

**Comment by letslookatcode** [comment]
I'm looking to switch to a new desktop environment, and after a lot of research I determined that Mate is my favorite. But no fractional scaling is a dealbreaker. I was trying to find a 1366x768 resolution 14" laptop like my old laptop (which would roughly match my 1920:1080 27" external screen in terms of DPI, adjusted for viewing distance), but they are no longer produced. Hence I need to do fractional scaling and per monitor. I couldn't care about any other feature or gimmick except this one. I think I'm not alone. Not sure why Linux desktops still don't support this nowadays essential feature, as HiDPI displays have been around for many years, and in fact non-HiDPI are hard to find. I hear it now works for Gnome, at least for native GTK applications. I hope for Gnome apps they use DPI aware applications, not actual after-the-fact scaling that would result in blurriness.

**Comment by lukefromdc** [comment]
Some time ago, I tested using xrandr to force different scaling oa secondary screen which was non-hidpi while using a 4K primary monitor. it worked, but was a bit sluggish and definately used more resources. This was on a big desktop machine at that

**Comment by lukefromdc** [comment]
Note that most of MATE from current git master supports running under wayland using any wlroots-based compositor. The panel(with applets built in-process), applets(if built in-process), caja, icons on the desktop etc all work, though the session is managed by the compositor not by MATE. At least in wayfire, both fractional scaling (floating point) and dissimilar scaling per monitor just plain work.

In some distros (notably Fedora) the volume control applet is buggy and has been for years. A crash of any in-process applet takes down the panel, so if running a wayland session in Fedora especially with removable or bluetooth sound devices it may pay to remove the volume control applet from the panel and add a launcher for mate-volume-control in it's place

**Comment by BrandowLucas** [comment]
+1

**Comment by lukefromdc** [comment]
Since the problem in the volume control applet (a weak pointer issue) has been found and fixed,
building with wayland support (applets in-process) is now much more practical. In wayland, at least
some compositors offer fractional scaling right now. Wayfire has a basic version of it, giving decimal
(float) scaling and that's getting further attention. Yes, I've tested that, and best of all you can even have 
dissimilar scaling on dissimilar monitors. In fact, anytime I use both of my pair of dissimilar monitors 
on wayland, I have the big one at 2x and the small one at 1x scale.

There is still a long way to go on scaling though, as scaling that does not align pixel boundaries with
those of the native widgets at unity scale can introduce blurring. Fixing that would require toolkits to 
handle this at the widget level rather than the compositor deal with it. Such toolkits would then have to
tell wayland compositors not to scale them again.

If Xrandr would catch up with wayland on scaling, we could have the equivalent of comnpositor level
scaling on Xorg too. Right fractional scaling on Xorg and GTK3 is as far as I know done by scaling
up by the numerator, then down by the denominator. Even dissimilar monitor resolution is heavy
and slow when done by Xrandr, acts like a GPU job is going to the CPU.
